src/openms/source/QC/Contaminants.cpp

```diff
@@ -35,113 +35,181 @@
+      else
       {
-        dropped_by_length += digestor.digest(AASequence::fromString(fe.sequence), current_digest);
+        digestor.setEnzyme(enzyme);
       }
-//      DigestedProtein temp;
-//      temp.id = fe.identifier;
-      for (auto const& s : current_digest)                                    //fill digested_db_ with digested sequences
+      //Get the missed cleavages for the digestor. If none are given, its default is 0.
+      UInt missed_cleavages(features.getProteinIdentifications()[0].getSearchParameters().missed_cleavages);
+      digestor.setMissedCleavages(missed_cleavages);
+
+      // stats for removing candidates
+      Size dropped_by_length(0);
+
+
+      //digest the contaminants database and add the peptides into the unordered set
+      for (const FASTAFile::FASTAEntry &fe : contaminants)
       {
-         digested_db_.insert(s.toUnmodifiedString());
-          //temp.peptides.push_back(s.toString());
+        vector<AASequence> current_digest;
+        //No digestion enzyme is given. Contaminant database will not get digested.
+        if (enzyme == "unknown_enzyme")
+        {
+          current_digest.push_back(AASequence::fromString(fe.sequence));
+        }
+        //digest the sequence
+        else
+        {
+          dropped_by_length += digestor.digest(AASequence::fromString(fe.sequence), current_digest);
+        }
+
+        //fill unordered set digested_db_ with digested sequences
+        for (auto const &s : current_digest)
+        {
+          digested_db_.insert(s.toUnmodifiedString());
+        }
       }
-      //digested_db_.push_back(temp);
     }
-  }
-  Int64 total;
-  Int64 cont;
-  double sum_total;
-  double sum_cont;
+    Int64 total = 0;
+    Int64 cont = 0;
+    double sum_total = 0.0;
+    double sum_cont = 0.0;
 
-  if (contaminants.empty())                                                    //no contamiants given
-    {
-      std::cerr << "FASTAFile is empty";
-      return;
-    }
 
-  for (auto& f : features)
-  {
-    if (f.getPeptideIdentifications().size() > 1)
+    //Check if peptides of featureMap are contaminants or not and add is_contaminant = 0/1 to the peptideidentification.
+    //If so, raise contaminants ratio.
+    for (auto &f : features)
     {
-      throw Exception::IllegalArgument(__FILE__,__LINE__,__FUNCTION__,"Too many peptideidentifications. Run IDConflictResolver first to remove ambiguities!");
+      if (f.getPeptideIdentifications().size() > 1)
+      {
+        throw Exception::IllegalArgument(__FILE__,
+                                         __LINE__,
+                                         __FUNCTION__,
+                                         "Too many peptideidentifications. Run IDConflictResolver first to remove ambiguities!");
+      }
+
+      // it exists one or zero Peptideidentifications in feature f
+      if (f.getPeptideIdentifications()[0].getHits().empty())
```

Comment:
check if feature has a PepId, i.e. if [0] is a valid operation (segfaults otherwise):
`if (f.getPeptideIdentifications().empty() || f.getPeptideIdentifications[0].getHits().empty())` continue;`